Hdf5 reader

jacobseiler opened the pull request:
@darrencroton This is a working version of HDF5 reader but has not been fully fleshed out.  I'm still working on it/cleaning things up.  Please do not merge yet.  I'll let you know when it gets the final @manodeep  tick of approval.

Commit:
Started work on HDF5 reader.

Need to check that reading of an array of attributes is working fine.
Since reading an array happens multiple times (for the data as well) maybe
should put in its own function.

Commit:
Reading in of table is correct.

Want to generalise so the user can specify the name of the fields of the
HDF5 file.

Commit:
Reworked directory structure.

Moved files dealing with io to their own directory with their own header
files.  In the final verison, core_io_tree.c should only act as a
wrapper file, calling the appropriate binary or HDF5 versions within io/
directory.

Commit:
Made io files more consistent with respective to local/external tags.

Commit:
Function that populates the metadata names works properly.

Have also added the -Werror option to the Makefile so I don't miss any
errors.

Commit:
Starting working on the reading of the actual tree.

Commit:
Hdf5 writer (#1)

* Added Simon hdf5 file writer

* Reading in of table is correct.

Want to generalise so the user can specify the name of the fields of the
HDF5 file.

* Committing all the changes. Need to add in the changes to core_io_tree and make those changes in sync with Jacobs reader section

* Fixing up compile errors

* Added object files and emacs backup files to gitigore

Commit:
Compiles and runs on g2 after merge.

Commit:
HDF5 reader now works fully.

The read in of the tree is done in an incredibly sloppy manner.  This
was done mainly because I wanted to get it done and ensured it works.
In future, this should be updated to include macros/other functions so
code doesn't need to be copy pasted.

The code has been checked by running with binary mode and HDF5 mode and
doing a diff of the outputs.  This resulted in no differences.  I'm
about to check back out master, run the base version and do a final
check to ensure nothing is different there.

Commit:
Reverted the accidental changes to allresults.

manodeep commented:
@jacobseiler Do the sage outputs agree on lhalotree binary input and hdf5 input from something like mini-mill?

Commit:
make clean removes the .exe. Claned up parameter file.

Commit:
make clean removes the .exe. Cleaned up parameter file.

jacobseiler commented:
@manodeep Yup. I made a small-script to convert Mini-Millennium binary to HDF5 (https://github.com/jacobseiler/short_scripts/tree/master/lhalo_to_hdf5).  The binary output of master and both modes of my branch (reading either binary or HDF5 input) produce bitwise-identical output.

Commit:
Merge branch 'HDF5_reader' of github.com:jacobseiler/sage into HDF5_reader

manodeep commented on code/core_allvars.h:
```diff
@@ -5,17 +5,26 @@
 #define  MAX_NODE_NAME_LEN 50
-#define  ABSOLUTEMAXSNAPS 1000
+#define  ABSOLUTEMAXSNAPS 1000  /* The largest number of snapshots for any simulation */
+#define  MAX_STRING_NAME_LEN  1024 /* Max length of a string containing a name*/
```
I had added `MAX_STRING_NAME_LEN` on the `hdf5_writer` branch but you can replace all instances of `MAX_STRING_NAME_LEN` with your define of `MAX_STRING_LEN`

manodeep commented on Makefile:
```diff
@@ -63,9 +84,9 @@ $(EXEC): $(OBJS)
 tidy:
 	rm -f $(OBJS) ./$(EXEC)
 
-all:  tidy $(EXEC) clean
```
I suspect that the reason why this `tidy` and `clean` existed is because the dependencies are not all captured in the Makefile. So, we should keep an eye on ensuring correct recompiling..

manodeep commented on code/core_allvars.c:
```diff
@@ -1,38 +1,52 @@
 int FoF_MaxGals;
-int Ntrees;			   // number of trees in current file 
-int NumGals;			 // Total number of galaxies stored for current tree 
+int Ntrees;			   /*  number of trees in current file  */
```
All of these were my bad. I am in the process of replacing all `//` style comments to `/* */` style comments.

manodeep commented on code/core_allvars.h:
```diff
@@ -185,24 +195,25 @@ struct halo_aux_data
 
-
-extern int    FirstFile;    // first and last file for processing 
+extern int    HDF5Output; 
```
The `HDF5Output` is not being used currently - right?

manodeep commented on code/core_allvars.h:
```diff
@@ -278,5 +294,27 @@ extern gsl_rng *random_generator;
+enum Valid_TreeTypes
+{
+  genesis_lhalo_hdf5 = 0,
+  lhalo_binary = 1
```
You can add a named field, something like `num_tree_types` field as the last enum, like so:

```C
enum Valid_TreeTypes
{
  genesis_lhalo_hdf5 = 0,
  lhalo_binary = 1,
  num_tree_types
}
```
Then, the `num_tree_types` field will auto-increment the last enum value by 1, i.e., `num_tree_types := 2`.
Afterwards, you can write input validation code like this:

```C
if(tree_type < 0 || tree_type >= num_tree_types) {
    /* print error message */
}
```

jacobseiler replied:
This causes a warning in my switch statements when I decide what tree reader to use ``code/core_io_tree.c:73: warning: enumeration value 'num_tree_types' not handled in switch``.  

I could obviously include the case explicitly in the switch but this seems redundant; I validate the user's value of ``TreeType`` in ``read_parameter_file()`` which is called before the switch.  There's no point in checking for ``case(num_tree_types)`` because we will error-exit before we hit the switch!

manodeep commented on code/core_io_tree.c:
```diff
@@ -11,42 +11,37 @@
-
   for(n = 0; n < NOUT; n++)
   {
+    TreeNgals[n] = mymalloc(sizeof(int) * Ntrees);
```
For the future, we should replace `mymalloc` to be something like `calloc`. Removes any potential overflows that can occur. For now, as long as we are compiling on 64bit platforms, `size_t` should be `uint64_t` and the overflow should not occur in the multiplication (since `Ntrees` gets promoted to `uint64_t` and the multiplication occurs between two `uint64_t`)

manodeep commented on code/core_io_tree.c:
```diff
@@ -74,46 +68,59 @@ void free_tree_table(void)
+}
 
-void load_tree(int filenr, int nr)
+void load_tree(int filenr, int treenr, enum Valid_TreeTypes my_TreeType)
```
Naming the var `treenr` makes it a lot clearer.

manodeep commented on code/core_proto.h:
```diff
@@ -10,14 +11,14 @@ int  join_galaxies_of_progenitors(int halonr, int nstart);
-void load_tree_table(int filenr);
-void load_tree(int filenr, int nr);
+void load_tree_table(int filenr, enum Valid_TreeTypes TreeType);
+void load_tree(int filenr, int nr, enum Valid_TreeTypes TreeType);
```
Should this `nr` be `treenr` as well?

manodeep commented on code/core_save.c:
```diff
@@ -23,67 +23,70 @@ void save_galaxies(int filenr, int tree)
+	{
+	  sprintf(buf, "%s/%s_z%1.3f_%d", OutputDir, FileNameGalaxies, ZZ[ListOutputSnaps[n]], filenr);
+	  
+	  if(!(save_fd[n] = fopen(buf, "r+")))
```
Please split this line into two; one for the `fopen`, and another with the check for `NULL`.

(Doing multiple things in a line is almost always a bad idea)

manodeep commented on code/core_save.c:
```diff
@@ -23,67 +23,70 @@ void save_galaxies(int filenr, int tree)
+	    }
+	  
+	  // write out placeholders for the header data.
+	  size_t size = (Ntrees + 2)*sizeof(int);
```
Why the `(Ntrees + 2)`? What are those 2 extra ints for? If those are some additional variables in the header, please write a comment saying what those two ints represent. And why they are ints (I am guessing format compatibility but I am not sure)

manodeep commented on code/core_save.c:
```diff
@@ -23,67 +23,70 @@ void save_galaxies(int filenr, int tree)
+	  size_t size = (Ntrees + 2)*sizeof(int);
+	  int* tmp_buf = (int*)malloc( size );
+	  memset( tmp_buf, 0, size );
+	  fwrite( tmp_buf, sizeof(int), Ntrees + 2, save_fd[n] );
```
Please check the output of `malloc` and `fwrite`;  unlike `free`, `malloc` and `fwrite` are not guaranteed. 

Well, `malloc` can not fail for `Too Small to Fail` memory requests, with the limit determined by by `PAGE_ALLOC_COSTLY_ORDER`. A few times the typical memory page size (4 KB) is should always succeed, otherwise the OS goes around killing random processes to reclaim memory (`Out of Memory Killer`, I am serious - that's the name). See [here](https://lwn.net/Articles/627419/) for details.

manodeep commented on code/io/io_save_hdf5.c:
```diff
@@ -0,0 +1,612 @@
+#include <stdio.h>
```
I am skipping this file because this is not being used currently - right?

jacobseiler replied:
Yup.

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+
+struct METADATA_NAMES 
+{
+char name_NTrees[MAX_STRING_LEN];
```
there's some whitespace issue with the field names.

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+#include "tree_hdf5.h"
+
+// Local Variables //
+hid_t hdf5_file;
```
Should this be declared `static` -- which means that this `hdf5_file` variable can only be accessed within this `tree_hdf5.c` file?

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+
+  struct METADATA_NAMES metadata_names;
+
+  sprintf(buf, "%s/%s.%d%s", SimulationDir, TreeName, filenr, TreeExtension);
```
Will you please replace `sprintf` with `snprintf`?

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+
+  sprintf(buf, "%s/%s.%d%s", SimulationDir, TreeName, filenr, TreeExtension);
+  hdf5_file = H5Fopen(buf, H5F_ACC_RDONLY, H5P_DEFAULT);
+  printf("%d\n", hdf5_file);
```
This seems like a debug print statement...

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+
+}
+
+#define READ_TREE_FIELD(file, name, type, h5type)\
```
Are you using this particular macro?

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+  /* Merger Tree Pointers */ 
+
+  // Descendant //
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/Descendant", treenr);
```
Excellent `snprintf` :)

manodeep replied:
I think you could wrap `snprintf` into a macro, because typing in `MAX_STRING_LEN - 1` might get annoying

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+
+  // Descendant //
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/Descendant", treenr);
+  buffer_int = malloc(sizeof(int32_t) * NHalos_ThisTree);
```
Check for successful malloc

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+
+  // Next Progenitor //
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/NextProgenitor", treenr);
+  buffer_int = malloc(sizeof(int32_t) * NHalos_ThisTree);
```
You can just have one giant buffer as `void *` allocated for the largest memory size you are going to read. And then cast to the appropriate type during each read.

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -0,0 +1,513 @@
+  free(buffer_int);
+
+  // FirstHaloInFOFgroup//
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/FirstHaloInFOFgroup", treenr);
```
There's definitely a convenience `macro` in your future :)

jacobseiler replied:
I think I'll forgo most of the suggested changes for this file as they'll probably change/be superseded by the ``macro``...

manodeep commented on output/allresults.py:
```diff
@@ -229,10 +232,10 @@ def StellarMassFunction(self, G):
-        ma = np.floor(max(mass)) + 2
-        NB = (ma - mi) / binwidth
-
+        mi = np.floor(min(mass)) - 2)
```
floating `)`?

manodeep commented on code/main.c:
```diff
@@ -163,11 +167,28 @@ int main(int argc, char **argv)
+  printf("Output 0 had %d galaxies\n", TotGalaxies[0]);
+
+/*
+  if(HDF5Output){
```
I am in two minds -- how hard will it be to finish off the hdf5 writer? If too hard, we should remove the traces from the existing codebase, (including `HDF5OUTPUT`..

jacobseiler replied:
In the scheme of things I don't think it will be too much; probably another day of work similar to the reader part.  However introducing the writer would also mean that ``allresults.py`` should be updated/an option added to handle HDF5 files.  I think that's where things get murky and we'd need to think about justifying the time investment.

Commit:
Cleaned up in response to Manodeep's comments.

Need to give a check over the entire code for validation of fwrites and
mallocs.  Actually I think the validation should just be put into
myfwrite and mymalloc and ensure that everything uses those functions.

Commit:
HDF5 reader now uses Macros.

Instead of copy pasting the code like a noob I've replaced it all with
two macros. I needed 2 because of fields such as Pos which has multiple
dimesions.

jacobseiler commented:
Latest commit has cleaned up ``tree_hdf5.c`` through the use of macros.  However the output of the three models (master, binary mode from my branch and hdf5 mode) are no longer bitwise identical. 

Using ``allresults.py`` I have printed out all the galaxy properties for each model and the galaxies have identical properties.  Unsure why they're not bitwise identical...

manodeep commented on code/io/tree_binary.c:
```diff
@@ -26,7 +26,7 @@ void load_tree_table_binary(int32_t filenr)
 
 	// open the file each time this function is called
-  snprintf(buf, "%s/%s.%d%s", MAX_STRING_LEN, SimulationDir, TreeName, filenr, TreeExtension);
+  snprintf(buf, MAX_STRING_LEN, "%s/%s.%d%s", SimulationDir, TreeName, filenr, TreeExtension);
```
Subtract 1 from the effective length

manodeep commented on code/io/tree_binary.c:
```diff
@@ -26,7 +26,7 @@ void load_tree_table_binary(int32_t filenr)
 	// open the file each time this function is called
-  snprintf(buf, "%s/%s.%d%s", MAX_STRING_LEN, SimulationDir, TreeName, filenr, TreeExtension);
+  snprintf(buf, MAX_STRING_LEN, "%s/%s.%d%s", SimulationDir, TreeName, filenr, TreeExtension);
   if(!(load_fd = fopen(buf, "r")))
```
Split into two statements

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -94,19 +94,50 @@ void load_tree_table_hdf5(int filenr)
+  double *buffer; // Buffer to hold the read HDF5 data.  
+                  // The largest data-type will be double. 
+
+  double *buffer_multipledim; // However also need a buffer three times as large to hold data such as position/velocity.
```
Just keep a single buffer -- allocate 3 times the memory ( or `dim` times) for the largest variable type read in

manodeep commented on code/io/tree_hdf5.c:
```diff
@@ -119,306 +150,65 @@ void load_tree_hdf5(int32_t filenr, int32_t treenr)
-  }
-  free(buffer_int);
+  /* Merger Tree Pointers */ 
+  READ_TREE_PROPERTY(Descendant, Descendant, 0, int);
```
Replace the `0` with `H5T_NATIVE_INT`

Commit:
Models are now properly bitwise identical.

I have yet to implement the comments that Manodeep made ~1 hr ago.

jacobseiler commented:
@manodeep Please help me to understand. Changing the length of ``char buf`` from ``MAX_STRING_LEN`` to ``1000`` resolved the bitwise issues; the outputs of all models are now identical. I have no clue why.

manodeep commented:
Hmm - that is strange. What happens if you change `char buf[1000]` back to `char buf[MAX_STRING_LEN];` but redefine `MAX_STRING_LEN` to be 1000. 

(This is after my initial reaction of **unpossible**)

manodeep commented:
And does the result change with the initial `char buf[1000]` if you run `sage` multiple times? And did you `diff -q` all pairs of files?

Commit:
Added default for switch cases.

Commit:
Initialized output_galaxy.

Comparison to Master (after initializing output_galaxy as well) is now
bitwise identical for both binary and HDF5 readers.  Valgrind produces
no memory leaks, warnings or errors.

darrencroton commented:
Just merged Initialize galaxy_output but the branch (now?) has conflicts. Can you have a look at where those come from and let me know?

manodeep committed:
Merge branch 'master' into HDF5_reader

manodeep commented:
Those were very strange conflicts! The conflicting lines in `main.c` appeared to be identical between the branches. 

@jacobseiler Will you confirm that what I did was okay

jacobseiler commented:
@manodeep I could not see the diff for the conflict-fix. 

@darrencroton Will you please open a new branch and then I can change the destination for this PR to this new branch. That way, we can avoid accidentally polluting the sage master branch.

manodeep commented:
@darrencroton Jacob and I have looked at the `diff` again -- looks fine now. We could not locate a couple of lines of code previously (from the conflict-fix) but now we see that in the full `diff`. Feel free to merge against master

darrencroton commented:
OK, done. All good. Thanks guys!
